master/buildbot/reporters/telegram.py

```diff
@@ -0,0 +1,1045 @@
+                if isinstance(data, dict):
+                    data = data.copy()
+                    text = data.pop('command')
+                    try:
+                        notify = data.pop('notify')
+                    except KeyError:
+                        notify = None
+                else:
+                    text, data, notify = data, {}, None
+            data['tquery'] = query
+            self.answer_query(query['id'], notify)
+            message = {
+                'from': query['from'],
+                'chat': original_message.get('chat'),
+                'text': text,
+            }
+            if 'reply_to_message' in original_message:
+                message['reply_to_message'] = original_message['reply_to_message']
+
+        chat = message['chat']
+
+        user = message.get('from')
+        if user is None:
+            self.log('No user in incoming message')
+            return defer.succeed('no user')
+
+        text = message.get('text')
+        if not text:
+            return defer.succeed('no text in the message')
+
+        contact = self.getContact(user=user, channel=chat)
+        data['tmessage'] = message
+        template, contact.template = contact.template, None
+        if text.startswith(self.commandPrefix):
+            d = contact.handleMessage(text, **data)
+        else:
+            if template:
+                text = template.format(shlex.quote(text))
+            d = contact.handleMessage(text, **data)
+        return d
+
+    @defer.inlineCallbacks
+    def _post(self, path, **kwargs):
+        try:
+            res = yield self.http_client.post(path, **kwargs)
+            ans = yield res.json()
+            if not ans.get('ok'):
+                raise ValueError("[{}] {}".format(res.code, ans.get('description')))
+        except AssertionError as err:
+            # just for tests
+            raise err
+        except Exception as err:
+            self.log("ERROR: cannot send Telegram request {}: {}".format(path, err))
+        else:
+            return ans.get('result', True)
+
+    @defer.inlineCallbacks
+    def _post_until_success(self, path, **kwargs):
+        logme = True
+        while True:
+            try:
+                res = yield self.http_client.post(path, **kwargs)
+                ans = yield res.json()
+                if not ans.get('ok'):
+                    self.log("ERROR: cannot send Telegram request {}: "
+                             "[{}] {}".format(path, res.code, ans.get('description')))
+                    return
+            except AssertionError as err:
+                # just for tests
+                raise err
+            except Exception as err:
+                msg = "ERROR: cannot send Telegram request {} (will try again): {}".format(path, err)
+                if logme:
+                    self.log(msg)
+                    logme = False
+                yield task.deferLater(self.reactor, self.retry_delay, lambda: None)
```

Comment:
There is asyncsleep in buildbot.util

http://docs.buildbot.net/current/developer/utils.html#buildbot.util.asyncSleep

Reply:
I don't know how to easily use `asyncSleep` with `TestReactor`... for unit test.

Reply:
Ok, I think I have managed to test it...